File: thesis/scripts/simulate_ddos.js
/**
 * simulate_ddos.js
 * Simulates a DDoS brute-force attack against the login endpoint.
 *
 * Usage:
 *   node scripts/simulate_ddos.js
 *
 * Requirements:
 *   - Server must be running: npm run dev
 */

const TARGET = 'http://localhost:3000/api/auth/login';
const EMAIL  = 'victim@example.com';

const PASSWORDS = [
  '123456', 'password', 'qwerty', 'admin', 'letmein',
  'welcome', 'monkey', 'dragon', 'master', '123123',
  'abc123', 'iloveyou', 'sunshine', 'princess', 'football',
  'shadow', 'superman', 'michael', 'password1', '12345678',
];

async function sendLoginRequest(password, index) {
  const start = Date.now();
  try {
    const res = await fetch(TARGET, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email: EMAIL, password }),
    });
    const elapsed = Date.now() - start;
    const data = await res.json();

    if (res.status === 429) {
      const retry = data.retryAfter ?? '?';
      console.log(`[${String(index).padStart(2, '0')}] BLOCKED  (429) — tried: '${password}' — retry in ${retry}s`);
      return 'blocked';
    } else {
      const label = res.status === 200 ? 'LOGIN OK' : 'Wrong pw';
      console.log(`[${String(index).padStart(2, '0')}] ${label} (${res.status}) — tried: '${password}' — ${elapsed}ms`);
      return 'allowed';
    }
  } catch (err) {
    console.log(`[${String(index).padStart(2, '0')}] ERROR — ${err.message}`);
    return 'error';
  }
}

async function main() {
  const total = PASSWORDS.length;
  console.log(`Starting DDoS simulation — ${total} requests to ${TARGET}`);
  console.log(`Target account: ${EMAIL}`);
  console.log('-'.repeat(60));

  let allowed = 0, blocked = 0, errors = 0;

  for (let i = 0; i < total; i++) {
    const result = await sendLoginRequest(PASSWORDS[i], i + 1);
    if (result === 'allowed') allowed++;
    else if (result === 'blocked') blocked++;
    else errors++;
  }

  console.log('-'.repeat(60));
  console.log(`RESULT : ${allowed} processed | ${blocked} blocked | ${errors} errors`);
  console.log(`Success: ${Math.round((allowed / total) * 100)}% of attack requests reached the server`);
  console.log(`Blocked: ${Math.round((blocked / total) * 100)}% of attack requests were rejected`);
}

main();

File: thesis/scripts/simulate_ddos.py
"""
simulate_ddos.py
Simulates a DDoS brute-force attack against the login endpoint.

Usage:
    python scripts/simulate_ddos.py

Requirements:
    pip install requests
    npm run dev  (server must be running on localhost:3000)
"""

import requests
import time

TARGET    = "http://localhost:3000/api/auth/login"
EMAIL     = "victim@example.com"
TOTAL     = 10

PASSWORDS = [
    "123456", "password", "qwerty", "admin", "letmein",
    "welcome", "monkey", "dragon", "master", "123123",
]

results = {"allowed": 0, "blocked": 0, "errors": 0}

print(f"Starting DDoS simulation — {TOTAL} requests to {TARGET}")
print(f"Target account: {EMAIL}")
print("-" * 60)

for i, password in enumerate(PASSWORDS[:TOTAL], start=1):
    try:
        start = time.time()
        r = requests.post(TARGET, json={"email": EMAIL, "password": password}, timeout=5)
        elapsed = round((time.time() - start) * 1000)

        if r.status_code == 429:
            results["blocked"] += 1
            retry = r.json().get("retryAfter", "?")
            print(f"[{i:02d}] BLOCKED  (429) — tried: '{password}' — retry in {retry}s")
        elif r.status_code in (200, 401):
            results["allowed"] += 1
            label = "LOGIN OK" if r.status_code == 200 else "Wrong pw"
            print(f"[{i:02d}] {label} ({r.status_code}) — tried: '{password}' — {elapsed}ms")
        else:
            results["errors"] += 1
            print(f"[{i:02d}] Unexpected ({r.status_code}) — {r.text[:80]}")

    except requests.exceptions.RequestException as e:
        results["errors"] += 1
        print(f"[{i:02d}] ERROR — {e}")

print("-" * 60)
print(f"RESULT : {results['allowed']} processed | {results['blocked']} blocked | {results['errors']} errors")
print(f"Success: {round(results['allowed'] / TOTAL * 100)}% of attack requests reached the server")
print(f"Blocked: {round(results['blocked'] / TOTAL * 100)}% of attack requests were rejected")
